File: 程序员面试金典/4.12/pathSum_2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# 评论区解法，速度快4倍
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        self.ans = 0
        prefix_sum = defaultdict(int)
        prefix_sum[0] = 1

        def dfs(node: TreeNode, total: int):
            if node is None:
                return
            total += node.val
            self.ans += prefix_sum[total-sum]
            prefix_sum[total] += 1
            dfs(node.left, total)
            dfs(node.right, total)
            prefix_sum[total] -= 1

        dfs(root, 0)
        return self.ans

File: 程序员面试金典/4.12/pathSum_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 使用散列表优化，速度提高5倍
class Solution {
    private Map<Integer, Integer> pathCount = new HashMap<>();

    public int pathSum(TreeNode root, int sum) {
        return countPathsWithSum(root, sum, 0);
    }

    private int countPathsWithSum(TreeNode root, int targetSum, int runningSum) {
        if (root == null)
            return 0;
        runningSum += root.val;
        int sum = runningSum - targetSum;
        int totalPaths = pathCount.getOrDefault(sum, 0);
        if (runningSum == targetSum)
            ++totalPaths;
        
        pathCount.put(runningSum, pathCount.getOrDefault(runningSum, 0) + 1);
        totalPaths += countPathsWithSum(root.left, targetSum, runningSum);
        totalPaths += countPathsWithSum(root.right, targetSum, runningSum);
        pathCount.put(runningSum, pathCount.getOrDefault(runningSum, 0) - 1);
        return totalPaths;
    }
}
